usb/device/targets/TARGET_STM/USBPhy_STM32.cpp

```diff
@@ -0,0 +1,522 @@
+    if (max_packet > MAXTRANSFER_SIZE) return false;
+    if (endpoint & 0x80) {
+        HAL_PCDEx_SetTxFiFo(&hpcd, endpoint & 0x7f, (max_packet / 4) + 1);
+        len = HAL_PCDEx_GetTxFiFo(&hpcd,endpoint & 0x7f);
+        MBED_ASSERT(len >= max_packet);
+    }
+    HAL_StatusTypeDef ret = HAL_PCD_EP_Open(&hpcd, endpoint, max_packet, type);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    return (ret == HAL_OK) ? true:false;
+}
+
+void USBPhyHw::endpoint_remove(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret = HAL_PCD_EP_Close(&hpcd, endpoint);
+    MBED_ASSERT(ret == HAL_OK);
+}
+
+void USBPhyHw::endpoint_stall(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret;
+    ret = HAL_PCD_EP_SetStall(&hpcd, endpoint);
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+void USBPhyHw::endpoint_unstall(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret;
+    ret = HAL_PCD_EP_ClrStall(&hpcd, endpoint);
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+bool USBPhyHw::endpoint_read(usb_ep_t endpoint, uint8_t *data, uint32_t size)
+{
+    _read_buf[EP_TO_LOG(endpoint)] = data;
+    _read_size[EP_TO_LOG(endpoint)] = size;
+
+    uint8_t* pBuf = (uint8_t *)pBufRx; //TODO - this buffer shouldn't be shared for multiple endpoints
+    HAL_StatusTypeDef ret;
+    // clean reception end flag before requesting reception
+    ret = HAL_PCD_EP_Receive(&hpcd, endpoint, pBuf, size);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    return true;
+}
+
+uint32_t USBPhyHw::endpoint_read_result(usb_ep_t endpoint)
+{
+    uint8_t *buffer =_read_buf[EP_TO_LOG(endpoint)];
+    uint8_t max_size = _read_size[EP_TO_LOG(endpoint)];
+
+    if (epComplete[EP_TO_IDX(endpoint)]==0) {
+        /*  no reception possible !!! */
+        return 0;
+    } else if ((epComplete[EP_TO_IDX(endpoint)]!=1)) {
+        return 0;
+    }
+    uint8_t  *buff = (uint8_t *)pBufRx;
+    uint32_t length = (uint32_t) HAL_PCD_EP_GetRxCount(&hpcd, endpoint);
+    memcpy(buffer, buff, length > max_size ? max_size : length);
+    epComplete[EP_TO_IDX(endpoint)]= 0;
+    return length;
+}
+
+bool USBPhyHw::endpoint_write(usb_ep_t endpoint, uint8_t *data, uint32_t size)
+{
+    HAL_StatusTypeDef ret;
+    // clean transmission end flag before requesting transmission
+    epComplete[EP_TO_IDX(endpoint)] = 2;
+    ret = HAL_PCD_EP_Transmit(&hpcd, endpoint, data, size);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    // update the status
+    if (ret != HAL_OK) return false;
+    // fix me return is too simple
+    return true;
+}
+
+void USBPhyHw::endpoint_abort(usb_ep_t endpoint)
+{
+    // TODO - stop the current transfer on this endpoint and don't call the IN or OUT callback
+}
+
+void USBPhyHw::process()
+{
+    HAL_PCD_IRQHandler(&instance->hpcd);
+    // Re-enable interrupt
+    NVIC_ClearPendingIRQ(USBHAL_IRQn);
+    NVIC_EnableIRQ(USBHAL_IRQn);
+}
+
+void USBPhyHw::_usbisr(void) {
+    NVIC_DisableIRQ(USBHAL_IRQn);
+    instance->events->start_process();
+}
+
+//TODO - remove this hack which allows HAL_Delay to work when interrupts are disabled
```

Comment:
The ST cube library implemented HAL_Delay by polling a counter that was incremented by a ticker. Since USB is protected by a critical section this interrupt never fired causing a hang.

Furthermore since this is in a critical section any substantial delay could have a serious impact on interrupt latency. Any ideas on how to prevent blocking?

Reply:
To solve this , the HAL_Delay used under critical section must  not rely on tick interruption, it should read a counter value till a counter timeout value is reached.

Reply:
The whole USBDevice class is protected with a critical section, so none of the calls can block. There were 3 places in the code with delays. Once I removed these then everything worked normally:
https://github.com/ARMmbed/mbed-os/blob/e43d21d4ef6025e94e4567961cb87ed2f0474a26/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L188
https://github.com/ARMmbed/mbed-os/blob/e43d21d4ef6025e94e4567961cb87ed2f0474a26/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L910
https://github.com/ARMmbed/mbed-os/blob/e43d21d4ef6025e94e4567961cb87ed2f0474a26/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L923

Are these delays required by the USB peripheral itself, or just to make things easier for the user? The bottom two enable / disable the USB pullup, so I would imagine this delay is not required by the peripheral, but instead required for disconnect / connect to be properly detected on USB.

Do you have any recommendations on how to proceed with these delays @jamike? Here are some possible paths forward:
- Add non-blocking versions of these same functions to the ST library
- Remove these delays from ST library code
- Lazily delay - only delay in the second of two calls for the time remaining. If the caller delays externally then no blocking is needed.

Reply:
I have also thinking to overwrite the HAL_Delay function (as it is declared weak) and instead call a function that will only read us_ticker_read(). Similar to PR #6538 but this has to be declared in a common area for all STM devices. But I don't know if it will solve your issue?

Reply:
Hi @bcostm, that would solve my immediate issue of a lockup. Unfortunately, it would violate timing requirements, as this code is running with interrupts disabled.